File: greenhouse_aiot/backend/models/crop.py
"""Crop model — maps to the `crops` table defined in D1 schema.sql."""

from datetime import datetime

from models import db


class Crop(db.Model):
    """Active crop batch planted in a specific zone."""

    __tablename__ = "crops"

    crop_id             = db.Column(db.Integer,     primary_key=True)
    zone_id             = db.Column(db.Integer,     db.ForeignKey("zones.zone_id"),           nullable=False)
    crop_type_id        = db.Column(db.Integer,     db.ForeignKey("crop_types.crop_type_id"), nullable=False)
    created_by          = db.Column(db.Integer,     db.ForeignKey("users.user_id"),           nullable=False)
    batch_code          = db.Column(db.String(50),  unique=True)
    quantity            = db.Column(db.Integer,     nullable=False)
    planted_at          = db.Column(db.Date,        nullable=False)
    expected_harvest_at = db.Column(db.Date)
    actual_harvest_at   = db.Column(db.Date)
    status              = db.Column(db.String(20),  nullable=False, default="germinating")
    notes               = db.Column(db.Text)
    created_at          = db.Column(db.DateTime,    default=datetime.utcnow)

    VALID_STATUSES: frozenset[str] = frozenset(
        {"germinating", "growing", "flowering", "harvested", "failed"}
    )

    def to_dict(self, include_zone: bool = False, include_crop_type: bool = False) -> dict:
        """Serialise to a JSON-safe dict."""
        data: dict = {
            "crop_id":             self.crop_id,
            "zone_id":             self.zone_id,
            "crop_type_id":        self.crop_type_id,
            "created_by":          self.created_by,
            "batch_code":          self.batch_code,
            "quantity":            self.quantity,
            "planted_at":          self.planted_at.isoformat() if self.planted_at else None,
            "expected_harvest_at": self.expected_harvest_at.isoformat() if self.expected_harvest_at else None,
            "actual_harvest_at":   self.actual_harvest_at.isoformat() if self.actual_harvest_at else None,
            "status":              self.status,
            "notes":               self.notes,
            "created_at":          self.created_at.isoformat() if self.created_at else None,
        }
        if include_zone and self.zone:
            data["zone"] = self.zone.to_dict()
        if include_crop_type and self.crop_type:
            data["crop_type"] = self.crop_type.to_dict()
        return data

    def __repr__(self) -> str:
        return f"<Crop {self.batch_code or self.crop_id} ({self.status})>"

File: greenhouse_aiot/spring-backend/src/main/java/com/aiot/greenhouse/model/Crop.java
package com.aiot.greenhouse.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Lote de cultivo activo plantado en una zona específica del invernadero.
 */
@Entity
@Table(name = "crops")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Crop {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "crop_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "zone_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler", "devices", "crops"})
    private Zone zone;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "crop_type_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    private CropType cropType;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "created_by", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler", "password", "authorities"})
    private User createdBy;

    @Column(name = "batch_code", unique = true, length = 50)
    private String batchCode;

    @Column(name = "quantity", nullable = false)
    private Integer quantity;

    @Column(name = "planted_at", nullable = false)
    private LocalDate plantedAt;

    @Column(name = "expected_harvest_at")
    private LocalDate expectedHarvestAt;

    @Column(name = "actual_harvest_at")
    private LocalDate actualHarvestAt;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20)
    @Builder.Default
    private CropStatus status = CropStatus.GERMINATING;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    /** Ciclo de vida de un lote de cultivo. */
    public enum CropStatus {
        GERMINATING, GROWING, FLOWERING, HARVESTED, FAILED
    }
}
